sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala

```diff
@@ -635,150 +644,84 @@ class ResolveSessionCatalog(
-          c.location, c.comment, c.external))
-
-      case (None, None) if !isV2Provider(conf.defaultDataSourceName) =>
-        Some(buildCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, conf.defaultDataSourceName,
-          c.options, c.location, c.comment, c.external))
-
-      case _ =>
-        None
-    }
-  }
-
-  private def buildV1Table(
-      ident: TableIdentifier,
-      c: CreateTableStatement): Option[CatalogTable] = {
-    val schema = c.tableSchema
-    (c.provider, c.serde) match {
-      case (Some(provider), Some(serde)) =>
-        throw new AnalysisException(
-          s"Cannot create table with both USING $provider and ${serde.describe}")
-
-      case (None, Some(serde)) =>
-        Some(buildHiveCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, serde, c.options, c.location,
-          c.comment, c.external))
-
-      case (None, None) =>
-        Some(buildHiveCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, SerdeInfo.empty, c.options,
-          c.location, c.comment, c.external))
-
-      case (Some(provider), None) if !isV2Provider(provider) =>
-        Some(buildCatalogTable(
-          ident, schema, c.partitioning, c.bucketSpec, c.properties, provider, c.options,
-          c.location, c.comment, c.external))
-
-      case _ =>
-        None
-    }
-  }
-
-  private def buildCatalogTable(
-      table: TableIdentifier,
-      schema: StructType,
-      partitioning: Seq[Transform],
-      bucketSpec: Option[BucketSpec],
-      properties: Map[String, String],
-      provider: String,
+  private def getStorageFormatAndProvider(
+      provider: Option[String],
       options: Map[String, String],
       location: Option[String],
-      comment: Option[String],
-      external: Boolean): CatalogTable = {
-    // Hive types are allowed if the provider is "hive"
-    if (!DDLUtils.isHiveTable(Some(provider))) {
-      assertNoCharTypeInSchema(schema)
-    }
-
-    if (external) {
-      throw new AnalysisException(s"Operation not allowed: CREATE EXTERNAL TABLE ... USING")
-    }
-
-    val storage = CatalogStorageFormat.empty.copy(
+      maybeSerdeInfo: Option[SerdeInfo],
+      ctas: Boolean): (CatalogStorageFormat, String) = {
+    val nonHiveStorageFormat = CatalogStorageFormat.empty.copy(
+      locationUri = location.map(CatalogUtils.stringToURI),
+      properties = options)
+    val defaultHiveStorage = HiveSerDe.getDefaultStorage(conf).copy(
       locationUri = location.map(CatalogUtils.stringToURI),
       properties = options)
 
-    val tableType = if (location.isDefined) {
-      CatalogTableType.EXTERNAL
+    if (provider.isDefined) {
+      // The parser guarantees that USING and STORED AS/ROW FORMAT won't co-exist.
+      assert(maybeSerdeInfo.isEmpty)
+      nonHiveStorageFormat -> provider.get
+    } else if (maybeSerdeInfo.isDefined) {
```

Comment:
The logic here is roughly merging 3 serde infos: the one user-specified, the one inferred from STORED AS, and the default one.